File: fsharp/sum-of-multiples/SumOfMultiples.fs
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//                     Version 2, December 2004
//
//  Copyright (C) 2016 Eric Bailey <eric@ericb.me>
//
//  Everyone is permitted to copy and distribute verbatim or modified
//  copies of this license document, and changing it is allowed as long
//  as the name is changed.
//
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//    TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
//
//   0. You just DO WHAT THE FUCK YOU WANT TO.
//

module SumOfMultiples

/// Return the sum of multiples of `d` in `ds` up to, but not including, `n`.
let sumOfMultiples ds n =
  let divides    n  = fun d -> n % d = 0
  let anyDivides ds = List.filter (fun n -> List.exists (divides n) ds)
  List.sum <| anyDivides ds [1..n - 1]

File: haskell/sum-of-multiples/SumOfMultiples.hs
-- |
-- Module      : SumOfMultiples
-- Copyright   : (c) Eric Bailey, 2015
-- License     : MIT
--
-- Maintainer  : Eric Bailey
-- Stability   : experimental
-- Portability : portable
--
-- Calculating the sum of multiples.
module SumOfMultiples where

-- | Given a number, returns the sum of all the multiples of 3 or 5
-- (@'sumOfMultiples' [3,5]@) up to but not including that number.
sumOfMultiplesDefault :: Integral a => a -> a
sumOfMultiplesDefault = sumOfMultiples [3, 5]

-- | Given a list of numbers and a number, returns the sum of all the multiples
-- of any of the numbers in the list up to but not including that number.
sumOfMultiples :: (Integral a) => [a] -> a -> a
sumOfMultiples = (sum .) . (. fromOneToExclusive) . anyDivides

-- | Given a number, @x@, returns the list of numbers from 1 to @x - 1@.
fromOneToExclusive :: Integral a => a -> [a]
fromOneToExclusive = enumFromTo 1 . pred

-- | Given two lists of numbers, @xs@ and @ys@, returns a possibly empty list
-- of every @y@ in @ys@ such that some @x@ in @xs@ 'divides' @y@.

-- |
-- __List comprehension__:
--
-- > anyDivides xs ys = [y | y <- ys, any (`divides` y) xs]
--
-- __LaTeX__:
--
-- @
-- \{ y : y \in Y \quad \land \quad \exists \, x \; ( x \in X \land x\ |\ y ) \;\}
-- @
anyDivides :: Integral a => [a] -> [a] -> [a]
anyDivides = filter . flip (any . flip divides)

-- | Given two numbers, returns @True@ if the first divides the second,
-- otherwise @False@.
divides :: Integral a => a -> a -> Bool
d `divides` n = n `rem` d == 0
